sr_robot_lib/src/MST.cpp

```diff
@@ -0,0 +1,140 @@
+/* Copyright 2021 Shadow Robot Company Ltd.
+*
+* This program is free software: you can redistribute it and/or modify it
+* under the terms of the GNU General Public License as published by the Free
+* Software Foundation version 2 of the License.
+*
+* This program is distributed in the hope that it will be useful, but WITHOUT
+* ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+* FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for
+* more details.
+*
+* You should have received a copy of the GNU General Public License along
+* with this program. If not, see <http://www.gnu.org/licenses/>.
+*
+*
+* @brief This is a class for accessing the data from the MST tactiles.
+*/
+
+#include "sr_robot_lib/MST.hpp"
+#include <sr_utilities/sr_math_utils.hpp>
+
+#include <memory>
+#include <string>
+#include <vector>
+
+#define TACTILE_DATA_LENGTH_BYTES TACTILE_DATA_LENGTH_BYTES_v2
+
+namespace tactiles
+{
+  template<class StatusType, class CommandType>
+  MST<StatusType, CommandType>::MST(ros::NodeHandle nh, std::string device_id,
+                                      std::vector<generic_updater::UpdateConfig> update_configs_vector,
+                                      operation_mode::device_update_state::DeviceUpdateState update_state,
+                                      boost::shared_ptr<std::vector<GenericTactileData> > init_tactiles_vector)
+          : GenericTactiles<StatusType, CommandType>(nh, device_id, update_configs_vector, update_state)
+  {
+    diagnostic_data = *init_tactiles_vector;
+    publisher = std::make_shared<ros::Publisher>(nh.advertise<sr_robot_msgs::MSTPalm>("mst", 1));
+  }
+
+  template<class StatusType, class CommandType>
+  int MST<StatusType, CommandType>::read12bits(char* buffer, int index)
+  {
+    int start = index * 3 / 2;
+    if (index % 2 == 0)
+    {
+      int byte1 = buffer[start];
+      int byte2 = buffer[start + 1];
+      return byte1 << 4 | byte2 >> 4 & 0x0F;
+    }
+    int byte1 = (int8_t)(buffer[start] << 4);
+    int byte2 = buffer[start + 1];
+    return byte1 << 4 | byte2 & 0xFF;
+  }
+
+  template<class StatusType, class CommandType>
+  void MST<StatusType, CommandType>::update(StatusType *status_data)
+  {
+    int tactile_mask = static_cast<int16u>(status_data->tactile_data_valid);
+    for (unsigned int id_sensor = 0; id_sensor < this->nb_tactiles; ++id_sensor)
+    {
+      switch (static_cast<int32u>(status_data->tactile_data_type))
+      {
+        case TACTILE_SENSOR_TYPE_MST_MAGNETIC_INDUCTION:
+          if (sr_math_utils::is_bit_mask_index_true(tactile_mask, id_sensor))
+          {
+            for (int i = 0; i < 7; i++)
```

Comment:
is 7 the number of sensors in an MST? can we place it in a variable for clarity?

Reply:
Fixed in b48a686520d3cadddc7d32a04be8cfdb92ead90d.